persistence/pg.go

```diff
@@ -260,68 +262,97 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
 	var settledHash *lntypes.Hash
 
 	err := p.conn.RunInTransaction(ctx, func(tx *pg.Tx) error {
-		// Look up the htlc hash.
-		hash, err := p.getHtlcHash(ctx, tx, key)
+		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
+		// htlcs for the same invoice have a status change concurrently.
+		invoice, err := p.selectInvoiceForUpdate(ctx, tx, key)
 		if err != nil {
 			return err
 		}
 
-		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
-		// htlcs for the same invoice are marked as settled concurrently.
-		var invoice dbInvoice
-		err = tx.ModelContext(ctx, &invoice).
-			Where("hash=?", hash).
-			For("UPDATE").
-			Select()
-		switch {
-		case err == pg.ErrNoRows:
-			return types.ErrInvoiceNotFound
+		now := time.Now().UTC()
 
-		case err != nil:
+		err = p.changeHTLCStatus(ctx, tx, key, types.StatusSettled, now)
+		if err != nil {
 			return err
 		}
 
-		now := time.Now().UTC()
-
-		htlc := dbHtlc{
-			Node:   key.Node,
-			ChanID: key.ChanID,
-			HtlcID: key.HtlcID,
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have been settled
```

Comment:
I am not sure if this is a real case, I could not find anything explicit in LND.

Reply:
I think this is real.